ceno_host/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+#![feature(trivial_bounds)]
+// TODO: create sp1 style host functionality.  Start with write and write_slice.
+
+use rkyv::{
+    Archive, Deserialize, Portable, Serialize,
+    api::high::{HighSerializer, HighValidator},
+    bytecheck::CheckBytes,
+    rancor::Error,
+    ser::allocator::ArenaHandle,
+    to_bytes,
+    util::AlignedVec,
+};
+
+#[derive(Archive, Deserialize, Serialize, Default)]
+pub struct CenoStdin {
+    pub items: Vec<Vec<u8>>,
+}
+
+impl CenoStdin {
+    pub fn write_slice(&mut self, item: AlignedVec) {
+        self.items.push(item.to_vec());
+    }
+    pub fn write(
+        &mut self,
+        item: &impl for<'a> Serialize<HighSerializer<AlignedVec, ArenaHandle<'a>, Error>>,
+    ) -> Result<(), Error> {
+        let bytes = to_bytes::<Error>(item)?;
+        self.items.push(bytes.to_vec());
+        Ok(())
+    }
+
+    pub fn finalise(&self) -> Result<AlignedVec, Error> {
+        to_bytes::<Error>(&self.items)
```

Comment:
serialization just happened once in the end, which imply iterating the data from begin to end.
can we do streaming processing for more CPU cache friendly?

Reply:
What goal are you trying to accomplish?

Serialisation happens in the host program running on native hardware.  That's plenty fast enough, even if we slowed it down 1000x times.  (And the total amount of data we want to serialise is limited by the fact that we need to deserialise that data in the VM, if it is to be useful.  So it will always be a relatively small amount of data.)

Reply:
Btw, the 'inner serialisation' happens one by one in something like a streaming fashion (both in the original and in the new design).  At the end we just collate all the individual items, which is mostly a bunch of mem-copies and a tiny bit of logic for the offsets.